Add handling for delayed message to redis transport

alexander-schranz opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | WIP
| Fixed tickets | Fixes #31711
| License       | MIT
| Doc PR        | symfony/symfony-docs#... TODO

Still in WIP: This pull request implements delayed messages for redis transport. It will park the messages in an own sorted set and if the time comes it will push the messages to the stream to make them available for all consumers. Because of a race condition when having multiple consumers it need to be checked if not accidently a message from the future is popped by zpopmin so the add function is called and there is check if the delay is in the present/past and only then add the message to the stream.

ragboyjr commented on src/Symfony/Component/Messenger/Transport/RedisExt/Connection.php:
```diff
@@ -93,6 +95,18 @@ public function get(): ?array
+        $queuedMessageCount = $this->connection->zcount($this->queue, 0, time());
+
+        if ($queuedMessageCount) {
+            foreach ($this->connection->zpopmin($this->queue, $queuedMessageCount) as $queuedMessage => $time) {
```
So, if there are 100 delayed messages, and only 5 delayed messages that actually are available now, wouldn't this end up cycling through the 100 delayed messages?

alexander-schranz replied:
@ragboyjr no `$queuedMessageCount` will be 5 then and so it only pop 5 messages.

ragboyjr replied:
Ah, understood.

Your technique for handling the race condition seems nice and simple, but one thing that concerns me would be loss of messages between the time that we pop it, then your redis instance has some failure, and then it fails when we try to re-add it (in the event of that race condition.

Part of me wonders if maybe, doing something where we pass in a LUA script to where it pops and pushes into the stream in one transaction would be preferable since the whole transaction would fail instead of part way through. What do you think?

alexander-schranz replied:
@ragboyjr As redis has only features to move a message from list to another list and not from a list to a stream. we would need to create a transaction then. never used that but I think it should be something like [this](https://github.com/phpredis/phpredis#transactions):

```php
    $this->connection->multi();
    foreach ($this->connection->zpopmin($this->queue, $queuedMessageCount) as $queuedMessage => $time) {
                $queuedMessage = json_encode($queuedMessage, true);
                // if a futured placed message is actually popped because of a race condition with
                // another running message consumer, the message is readded to the queue by add function
                // else its just added stream and will be available for all stream consumers
                $this->add($queuedMessage['body'], $queuedMessage['headers'], (time() - $time) * 1000);
    }

    $this->connection->exec();
```

we only need to start the transaction if there are really queued messages so its not blocking if there are none.

ragboyjr replied:
Y, I think a redis transaction would work as well, would like to see if that could be something we could cover in an integration test of adding items to the queue, then having redis fail and making sure we don't lose any messages.

alexander-schranz replied:
Transactions don't work as you can't read inside a transaction.
So to only solution would be to park this messages in another list as every consumer need its own name we postfix the new list with stream + group + consumer names and so its a unique list only for this consumer to avoid a conflicts with other running consumers. If it crash the consumer need to check if he has still things in this list to push to the stream.

weaverryan commented:
Ping @alexander-schranz! It looks like this is technically challenging. Is the path clear? Or are we stuck on those details?

alexander-schranz commented:
@weaverryan theoretically there is a little time where a message could get lost when the consumer crashes. Between read from the queued/delayed message sorted set list and adding to the stream.

https://github.com/symfony/symfony/blob/7407abd8e5ef3672075bb897bb7518195017a194/src/Symfony/Component/Messenger/Transport/RedisExt/Connection.php#L120-L128

I did now change that only one message is read and added to the stream.

ragboyjr commented:
> theoretically there is a little time where a message could get lost when the consumer crashes.

@alexander-schranz do you think utilizing a lua script similar to laravel would be able to prevent this issue?

ragboyjr commented:
@alexander-schranz mind if I take a stab at implementing the transactions part as a lua script?

ragboyjr commented:
I have a bit of time this weekend, and I'd love to see this feature for the redis adapter.

alexander-schranz commented:
@ragboyjr sure. Basically its the `zpopmin` -> `zadd` / `xadd` which we need to move into a transaction. Thank you to have a look at it I'm currently short in time.

Mrkisha commented:
Is there any progress on this PR?

alexander-schranz commented:
@Mrkisha rebased and fixed the current implemention and added some tests. So whats currently is not handled is when the redis crash between moving a delayed message from the delayed queue to the stream it could get lost.

The travis CI seems to have some changes not sure why the tests can not longer connect to the redis.

Mrkisha commented:
@alexander-schranz I think you're doing great job. You're almost there. Probably some minor fixes to satisfy travis-ci.

Mrkisha commented:
@alexander-schranz can I somehow help you out so we speed the whole process for this feature?

alexander-schranz commented:
@Mrkisha I tried to change the url to the one which was given in the error. If that still don't work it would be great if you could have a look at the Travis CI problem.

alexander-schranz commented:
@Mrkisha for questions you can also reach me in the [Symfony Slack Chat](https://symfony.com/community) under `@alex_s_`

chalasr commented on src/Symfony/Component/Messenger/Tests/Transport/RedisExt/RedisExtIntegrationTest.php:
```diff
@@ -55,11 +55,48 @@ public function testGetTheFirstAvailableMessage()
+
+        $this->connection->add($body, $headers, 500);
+        $this->connection->add($body, $headers, 500);
+        usleep(500000);
```
We should mark the test class as `@group time-sensitive`

alexander-schranz replied:
Added 👍

chalasr commented on src/Symfony/Component/Messenger/Transport/RedisExt/Connection.php:
```diff
@@ -124,6 +126,24 @@ public function get(): ?array
             $this->setup();
         }
 
+        $queuedMessageCount = $this->connection->zcount($this->queue, 0, $this->getCurrentTimeInMilliseconds());
```
do we need some error handling here? (same for the loop below)

alexander-schranz replied:
Added try catch for it. Thank you for review 👍

alexander-schranz commented on src/Symfony/Component/Messenger/Transport/RedisExt/Connection.php:
```diff
@@ -246,4 +285,18 @@ public function setup(): void
+
+    /**
+     * @internal
+     */
```
Did remove it.

alexander-schranz commented:
@sroze do you want to have a look at this :)?

weaverryan commented:
@alexander-schranz Is this done & ready? Or are there still some issues?

alexander-schranz commented:
@weaverryan ready from my side :)

chalasr commented:
I'm playing with this right now, will try to leave a review tonight.

Tobion commented on src/Symfony/Component/Messenger/Transport/RedisExt/Connection.php:
```diff
@@ -203,24 +233,43 @@ public function reject(string $id): void
 
-    public function add(string $body, array $headers): void
+    /**
+     * @param int $delay The delay in milliseconds
```
please call the param `$delayInMs` like in amqp connection. then you don't need this phpdoc

alexander-schranz replied:
@Tobion thank you for your review. Did change this!

Tobion commented on src/Symfony/Component/Messenger/Transport/RedisExt/Connection.php:
```diff
@@ -246,4 +295,15 @@ public function setup(): void
+        return (int) (microtime(true) * 1000);
+    }
+
+    public function cleanup()
```
please mark as internal and add void return like the other methods have

alexander-schranz replied:
Did remove the `@internal` after the review of @chalasr see https://github.com/symfony/symfony/pull/31977#discussion_r334295302 because the whole class is marked as internal.
Should I still add it?

/cc @chalasr

Tobion replied:
Oh I didn't know. Then it's fine.

Tobion reviewed (changes requested):
please add changelog entry

alexander-schranz committed:
Add handling for delayed message to redis transport

chalasr commented:
Thank you @alexander-schranz.

ragboyjr commented:
Holy Wow!!! so excited to se this merged in!!!

Mrkisha commented:
Awesome!
@alexander-schranz  Great work!
